### ISSUE TITLE
"Class" is compatible with "Self@Class"

### ISSUE THREAD

**Issue by juk-r** [microsoft/pyright#6261]
The following  example produce no errors:
```python
from typing import Self

class A:
    def self_arg(self, other: Self):
       ...

    def call_self_arg(self):
        self.self_arg(A()) # ok, must be error

    def get_instance(self) -> Self:
        return A() # ok, must be error
```

**Comment by Viicos** [comment]
I think this is working as expected: `typing.Self` doesn't really mean "the current tied instance", but rather some sort of type variable bound to the encapsulating class (quoted from PEP 673). Because it is tied to the class, it shouldn't have any relation to the current instance. One example illustrating this:

```python
class A:
    def return_self(self) -> Self:
        return self

a = A()
a.attr = 1
b = a.return_self()
b.attr  # Type checkers shouldn't allow this
```

**Comment by erictraut** [comment]
The OP is correct. This is a bug (false negative). Pyright should generate an error here. You can think of `Self` as a type variable that has an upper bound of the current class. 

You can see how the comparable code using an actual TypeVar produces the correct errors.

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIAUDQMpQC8hJFIAFAES0cA0UARmACuKACYMOAIQ4BKGgGMANuQDOKqJIBcVKLqijiwKCuKLgAfUpo2Js5qi0%2BYGAAtiIe7Rna9vqADpAml8DI3lyRUVzWwsrG1NgT28dP10Y-xjLEGtJNhkZXQBiZgRiKHdwahTdUKg0YhhzVBUYchR5Yni7BwKAWgA%2BBx9UqBAGoRAUDTyikrKK3CogA)

```python
from typing import TypeVar

S = TypeVar("S", bound="B")

class B:
    def self_arg(self: S, other: S):
        ...

    def call_self_arg(self: S):
        self.self_arg(B())  # Type error

    def get_instance(self: S) -> S:
        return B()  # Type error
```

Interestingly, both mypy and pyre (two other Python type checkers) correctly produce the second of two errors that should be produced in the OP's example, but they don't produce the first error. In other words, they also have a bug here.

**Comment by Viicos** [comment]
Ok so if I understand correctly, the example the OP provided is an issue because he is using the "hardcoded" class name, which means it will not play nicely with subclasses? In other terms, should the following example work as expected?

```python
from typing import Self

class A:
    def self_arg(self, other: Self):
       ...

    def call_self_arg(self):
        self.self_arg(type(self)())

    def get_instance(self) -> Self:
        return type(self)()
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.335, which I just published. It will also be included in a future release of pylance.
